src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/DotNetCoreProjectCompatibilityDetector.cs

```diff
@@ -0,0 +1,317 @@
+                        await TaskScheduler.Default;
+
+                        // We need to check if this project has been newly created. Our projects will implement IProjectCreationState -we can 
+                        // skip any that don't
+                        var projectCreationState = project.Services.ExportProvider.GetExportedValueOrDefault<IProjectCreationState>();
+                        if(projectCreationState != null && !projectCreationState.WasNewlyCreated)
+                        {
+                            CompatibilityLevel compatLevel = await GetProjectCompatibilityAsync(project).ConfigureAwait(false);
+                            if(compatLevel != CompatibilityLevel.Supported)
+                            {
+                                await WarnUserOfiIncompatibleProjectAsync(compatLevel).ConfigureAwait(false);
+                            }
+                        }
+                    });
+                }
+            }
+
+            return VSConstants.S_OK;
+        }
+
+        public int OnAfterCloseSolution(object pUnkReserved)
+        {
+            // Clear state flags
+            _compatibilityLevelWarnedForThisSolution = CompatibilityLevel.Supported;
+            _solutionOpened = false;
+            return VSConstants.S_OK;
+        }
+
+        /// <summary>
+        /// Fired when the solution load process is fully complete, including all background loading 
+        /// of projects. This event always fires after the initial opening of a solution 
+        /// </summary>
+        public int OnAfterBackgroundSolutionLoadComplete()
+        {
+            // Schedule this to run on idle
+            ThreadHandling.Value.JoinableTaskFactory.StartOnIdle(async () =>
+            {
+                // Run on the background
+                await TaskScheduler.Default;
+
+                CompatibilityLevel finalCompatLevel = CompatibilityLevel.Supported;
+                IProjectService projectService = ProjectServiceAccessor.Value.GetProjectService();
+                IEnumerable<UnconfiguredProject> projects = projectService.LoadedUnconfiguredProjects;
+                foreach (var project in projects)
+                {
+                    // Track the most severe compatibility level
+                    CompatibilityLevel compatLevel = await GetProjectCompatibilityAsync(project).ConfigureAwait(false);
+                    if(compatLevel != CompatibilityLevel.Supported && compatLevel > finalCompatLevel)
+                    {
+                        finalCompatLevel = compatLevel;
+                    }
+                }
+            
+                if (finalCompatLevel != CompatibilityLevel.Supported)
+                {
+                    // Warn the user.
+                    await WarnUserOfiIncompatibleProjectAsync(finalCompatLevel).ConfigureAwait(false);
+               }
+
+               // Used to know when to process newly added projects
+               _solutionOpened = true;
```

Comment:
What happens if there is a solution loaded when the package loads (can happen if loaded for Tools\Options, etc)?

Reply:
The code will be initialized and hook into events but it won't do anything until the solution closes and a new one is opened. Generally this is OK, because it means the solution doesn't contain any .netcore projects. However. it would miss the case of a customer adding an existing .NET core project to the already open solution. I'll address that case